### PR TITLE
[SPARK-9740] [SPARK-9592] [SPARK-9210] [SQL] Change the default behavior of First/Last to RESPECT NULLS.

### PR DESCRIPTION
**Pull request by yhuai**
I am changing the default behavior of `First`/`Last` to respect null values (the SQL standard default behavior).

https://issues.apache.org/jira/browse/SPARK-9740

### CODE REVIEW

**Comment by yhuai** [comment]
@hvanhovell I am making the default in this PR.

**Comment by hvanhovell** [comment]
LGTM. One final question, shouldn't we introduce a ```skipNulls``` parameter? Or do you want to address this in a follow-up PR?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala [hunk 1]
```diff
@@ -130,23 +138,36 @@ case class First(child: Expression) extends AlgebraicAggregate {
...
   override val updateExpressions = Seq(
-    /* first = */ If(IsNull(first), child, first)
+    /* first = */ If(valueSet, first, child),
+    /* valueSet = */ If(valueSet, valueSet, Literal.create(true, BooleanType))
   )
 
   override val mergeExpressions = Seq(
-    /* first = */ If(IsNull(first.left), first.right, first.left)
+    /* first = */ If(valueSet, first.left, first.right),
+    /* valueSet = */ If(valueSet, valueSet, Literal.create(true, BooleanType))
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala]: We could for now simplify this to ```Literal.create(true, BooleanType)```.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala [hunk 2]
```diff
@@ -130,23 +138,36 @@ case class First(child: Expression) extends AlgebraicAggregate {
...
   override val initialValues = Seq(
-    /* first = */ Literal.create(null, child.dataType)
+    /* first = */ Literal.create(null, child.dataType),
+    /* valueSet = */ Literal.create(false, BooleanType)
   )
 
   override val updateExpressions = Seq(
-    /* first = */ If(IsNull(first), child, first)
+    /* first = */ If(valueSet, first, child),
+    /* valueSet = */ If(valueSet, valueSet, Literal.create(true, BooleanType))
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala]: We could for now simplify this to ```Literal.create(true, BooleanType)```.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala (hunk 1 above)
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala]: ```Or(valueSet.left, valueSet.right)```?

**Comment by yhuai** [comment]
yeah. I think that only require a small change. Let me add it.

**Comment by hvanhovell** [comment]
Besides the ```valueSet``` update/merge expressions LGTM.

**Comment by yhuai** [comment]
@hvanhovell Alright. I added the `ignoreNulls` flag.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregates.scala [hunk 3]
```diff
@@ -630,59 +632,114 @@ case class CombineSetsAndSumFunction(
...
   var result: Any = null
 
   override def update(input: InternalRow): Unit = {
-    result = input
+    val value = expr.eval(input)
+    if (ignoreNulls && value != null) {
+      result = value
+    } else {
+      result = value
+    }
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregates.scala]: @ggupta81 This is the fix of SPARK-9592 for our master and 1.5 branch.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregates.scala [hunk 4]
```diff
@@ -630,59 +632,114 @@ case class CombineSetsAndSumFunction(
...
+  def this() = this(null, null.asInstanceOf[Boolean], null) // Required for serialization.
 
   var result: Any = null
 
   override def update(input: InternalRow): Unit = {
-    result = input
+    val value = expr.eval(input)
+    if (ignoreNulls && value != null) {
+      result = value
+    } else {
```
Comment (ggupta81) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregates.scala]: Both if and else and executing the same code.

Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregates.scala]: oops. Will fix it.

**Comment by ggupta81** [comment]
Should i close my original pull request now?
https://github.com/apache/spark/pull/7928

On Wed, Aug 12, 2015 at 10:38 AM, UCB AMPLab <notifications@github.com>
wrote:

> Merged build started.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/8113#issuecomment-130167350>.
>



-- 

*Gaurav Gupta*Engineering Manager @ Adobe
9971666780

**Comment by yhuai** [comment]
@ggupta81 Yeah. We can close #7928. Can you open a pr to fix branch 1.4, or reopen #7233 and update that? Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala [hunk 5]
```diff
@@ -130,24 +147,67 @@ case class First(child: Expression) extends AlgebraicAggregate {
...
 
-  override val updateExpressions = Seq(
-    /* first = */ If(IsNull(first), child, first)
-  )
+  override val updateExpressions = {
+    val litTrue = Literal.create(true, BooleanType)
+    if (ignoreNulls) {
+      Seq(
+        /* first = */ If(Or(valueSet, IsNull(child)), first, child),
+        /* valueSet = */ If(Or(valueSet, IsNull(child)), valueSet, litTrue)
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala]: Nit: ```Or(valueSet, Not(IsNull(child)))``` is a bit shorter.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala [hunk 6]
```diff
@@ -130,24 +147,67 @@ case class First(child: Expression) extends AlgebraicAggregate {
...
+    val litTrue = Literal.create(true, BooleanType)
+    if (ignoreNulls) {
+      Seq(
+        /* first = */ If(Or(valueSet.left, IsNull(first.right)), first.left, first.right),
+        /* valueSet = */ If(Or(valueSet.left, IsNull(first.right)), valueSet.left, litTrue)
+      )
+    } else {
+      Seq(
+        /* first = */ If(valueSet.left, first.left, first.right),
+        /* valueSet = */ litTrue
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala]: Is it possible that two non-updated aggregates get merged? If it is, then this should be: ```Or(valueSet.left. valueSet.right)```

**Comment by hvanhovell** [comment]
One more small thing. We should probably also add the ```ignoreNulls``` option to the ```first``` and ```last``` dataframe functions.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala [hunk 7]
```diff
@@ -130,24 +147,67 @@ case class First(child: Expression) extends AlgebraicAggregate {
...
 
-  override val mergeExpressions = Seq(
-    /* first = */ If(IsNull(first.left), first.right, first.left)
-  )
+  override val mergeExpressions = {
+    val litTrue = Literal.create(true, BooleanType)
+    if (ignoreNulls) {
+      Seq(
+        /* first = */ If(Or(valueSet.left, IsNull(first.right)), first.left, first.right),
+        /* valueSet = */ If(Or(valueSet.left, IsNull(first.right)), valueSet.left, litTrue)
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala]: Nit: ```Or(valueSet.left. valueSet.right)``` is shorter.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala [hunk 8]
```diff
@@ -130,24 +147,67 @@ case class First(child: Expression) extends AlgebraicAggregate {
...
+  }
 
-  override val mergeExpressions = Seq(
-    /* first = */ If(IsNull(first.left), first.right, first.left)
-  )
+  override val mergeExpressions = {
+    val litTrue = Literal.create(true, BooleanType)
+    if (ignoreNulls) {
+      Seq(
+        /* first = */ If(Or(valueSet.left, IsNull(first.right)), first.left, first.right),
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala]: I don't think we need to check ```IsNull(first.right)```.

On a more general note: I think we don't need to separate the ```ignoreNulls = false``` and ```ignoreNulls = true``` paths. This should be fine for both:
```
/* first = */ If(valueSet.left, first.left, first.right),
/* valueSet = */ Or(valueSet.left, valueSet.right)
```

**Comment by yhuai** [comment]
@hvanhovell Since we already pass the feature freeze deadline, I will not add new interfaces to DataFrame API (this PR is mainly about fixing the default behavior). If users request adding DF functions that expose `ignoreNulls` flag, we can do it in 1.6.

**Comment by yhuai** [comment]
test this please

**Comment by marmbrus** [comment]
Thanks!  Merging to master.

**Pull request closed without merging**